x/leverage/keeper/msg_server_test.go

```diff
@@ -172,3 +173,1545 @@ func (s *IntegrationTestSuite) TestUpdateRegistry() {
+		uToken               sdk.Coin
+		expectFromBalance    sdk.Coins
+		expectFromCollateral sdk.Coins
+		expectedTokens       sdk.Coin
+		err                  error
+	}
+
+	app, ctx, srv, require := s.app, s.ctx, s.msgSrvr, s.Require()
+
+	// create and fund a supplier with 100 UMEE and 100 ATOM, then supply 100 UMEE and 50 ATOM
+	// also collateralize 75 of supplied UMEE
+	supplier := s.newAccount(coin(umeeDenom, 100_000000), coin(atomDenom, 100_000000))
+	s.supply(supplier, coin(umeeDenom, 100_000000))
+	s.collateralize(supplier, coin("u/"+umeeDenom, 75_000000))
+	s.supply(supplier, coin(atomDenom, 50_000000))
+
+	// create and modify a borrower to force the uToken exchange rate of ATOM from 1 to 1.2
+	borrower := s.newAccount(coin(atomDenom, 100_000000))
+	s.supply(borrower, coin(atomDenom, 100_000000))
+	s.collateralize(borrower, coin("u/"+atomDenom, 100_000000))
+	s.borrow(borrower, coin(atomDenom, 10_000000))
+	s.tk.SetBorrow(ctx, borrower, coin(atomDenom, 40_000000))
+
+	// create an additional supplier (UMEE, DUMP, PUMP tokens)
+	other := s.newAccount(coin(umeeDenom, 100_000000), coin(dumpDenom, 100_000000), coin(pumpDenom, 100_000000))
+	s.supply(other, coin(umeeDenom, 100_000000))
+	s.supply(other, coin(pumpDenom, 100_000000))
+	s.supply(other, coin(dumpDenom, 100_000000))
+
+	// create a DUMP (historic price 1.00, current price 0.50) borrower
+	// using PUMP (historic price 1.00, current price 2.00) collateral
+	dumpborrower := s.newAccount(coin(pumpDenom, 100_000000))
+	s.supply(dumpborrower, coin(pumpDenom, 100_000000))
+	s.collateralize(dumpborrower, coin("u/"+pumpDenom, 100_000000))
+	s.borrow(dumpborrower, coin(dumpDenom, 20_000000))
+	// collateral value is $200 (current) or $100 (historic)
+	// borrowed value is $10 (current) or $20 (historic)
+	// collateral weights are always 0.25 in testing
+
+	// create a PUMP (historic price 1.00, current price 2.00) borrower
+	// using DUMP (historic price 1.00, current price 0.50) collateral
+	pumpborrower := s.newAccount(coin(dumpDenom, 100_000000))
+	s.supply(pumpborrower, coin(dumpDenom, 100_000000))
+	s.collateralize(pumpborrower, coin("u/"+dumpDenom, 100_000000))
+	s.borrow(pumpborrower, coin(pumpDenom, 5_000000))
+	// collateral value is $50 (current) or $100 (historic)
+	// borrowed value is $10 (current) or $5 (historic)
+	// collateral weights are always 0.25 in testing
+
+	tcs := []testCase{
+		{
+			"unregistered base token",
+			supplier,
+			coin("abcd", 80_000000),
+			nil,
+			nil,
+			sdk.Coin{},
+			types.ErrNotUToken,
+		},
+		{
+			"base token",
+			supplier,
+			coin(umeeDenom, 80_000000),
+			nil,
+			nil,
+			sdk.Coin{},
+			types.ErrNotUToken,
+		},
+		{
+			"insufficient uTokens",
+			supplier,
+			coin("u/"+umeeDenom, 120_000000),
+			nil,
+			nil,
+			sdk.Coin{},
+			types.ErrInsufficientBalance,
+		},
+		{
+			"withdraw from balance",
+			supplier,
+			coin("u/"+umeeDenom, 10_000000),
+			sdk.NewCoins(coin("u/"+umeeDenom, 10_000000)),
```

Comment:
let's create a helper function `ucoin` which will do the `"u/"+`, and panic if denom is already a utoken,